pkg/resources/management.cattle.io/v3/projectroletemplatebinding/validator.go

```diff
@@ -152,50 +142,65 @@ func clusterFromProject(project string) (string, string) {
 }
 
 // validUpdateFields checks if the fields being changed are valid update fields.
-func validateUpdateFields(oldPRTB, newPRTB *apisv3.ProjectRoleTemplateBinding) error {
-	var invalidFieldName string
+func validateUpdateFields(oldPRTB, newPRTB *apisv3.ProjectRoleTemplateBinding, fieldPath *field.Path) *field.Error {
+	const reason = "field is immutable"
 	switch {
 	case oldPRTB.RoleTemplateName != newPRTB.RoleTemplateName:
-		invalidFieldName = "referenced roleTemplate"
+		return field.Invalid(fieldPath.Child("roleTemplateName"), newPRTB.RoleTemplateName, reason)
 	case oldPRTB.ProjectName != newPRTB.ProjectName:
-		invalidFieldName = "projectName"
+		return field.Invalid(fieldPath.Child("projectName"), newPRTB.ProjectName, reason)
 	case oldPRTB.UserName != newPRTB.UserName && oldPRTB.UserName != "":
-		invalidFieldName = "userName"
+		return field.Invalid(fieldPath.Child("userName"), newPRTB.UserName, reason)
 	case oldPRTB.UserPrincipalName != newPRTB.UserPrincipalName && oldPRTB.UserPrincipalName != "":
-		invalidFieldName = "userPrincipalName"
+		return field.Invalid(fieldPath.Child("userPrincipalName"), newPRTB.UserPrincipalName, reason)
 	case oldPRTB.GroupName != newPRTB.GroupName && oldPRTB.GroupName != "":
-		invalidFieldName = "groupName"
+		return field.Invalid(fieldPath.Child("groupName"), newPRTB.GroupName, reason)
 	case oldPRTB.GroupPrincipalName != newPRTB.GroupPrincipalName && oldPRTB.GroupPrincipalName != "":
-		invalidFieldName = "groupPrincipalName"
+		return field.Invalid(fieldPath.Child("groupPrincipalName"), newPRTB.GroupPrincipalName, reason)
 	case (newPRTB.GroupName != "" || oldPRTB.GroupPrincipalName != "") && (newPRTB.UserName != "" || oldPRTB.UserPrincipalName != ""):
-		invalidFieldName = "both user and group"
+		return field.Forbidden(fieldPath,
+			"binding must target either a user [userName]/[userPrincipalName] OR a group [groupName]/[groupPrincipalName]")
+	case newPRTB.ServiceAccount != "":
+		return field.Forbidden(fieldPath.Child("serviceAccount"), "update is not allowed")
 	default:
 		return nil
 	}
-
-	return fmt.Errorf("cannot update %s for clusterRoleTemplateBinding %s: %w", invalidFieldName, oldPRTB.Name, admission.ErrInvalidRequest)
 }
 
 // validateCreateFields checks if all required fields are present and valid.
-func (a *admitter) validateCreateFields(newPRTB *apisv3.ProjectRoleTemplateBinding) error {
+func (a *admitter) validateCreateFields(newPRTB *apisv3.ProjectRoleTemplateBinding, fieldPath *field.Path) error {
+	if newPRTB.ServiceAccount != "" {
+		return field.Forbidden(fieldPath.Child("serviceAccount"), "")
+	}
```

Comment:
Why can't this be empty, and what is it being used for today?

Reply:
It's got a Norman tag: `norman:"nocreate,noupdate"`, so I am just emulating this here. Let me see what value it normally has with a real setup...

Reply:
In a PRTB for a user, that `serviceAccount` field is missing. I am seeing only a couple of places where the property is read. 
https://github.com/maxsokolovsky/rancher/blob/96f5ed3c4faefa8580ab9a5ceae204c74c058f86/pkg/controllers/managementuserlegacy/monitoring/projectHandler.go#L351
Here, for example, it's used in managementuserlegacy as an indexer func.

I am not sure if this field is really used anymore.

Reply:
Maybe it used to be important for projects? @cmurphy, do you happen to know?

Reply:
> Why can't this be empty, and what is it being used for today?

@KevinJoiner I think it's saying the opposite, it's saying it has to be empty (so it's useless)

> Maybe it used to be important for projects?

It's so you can create a binding with a service account as a subject: https://github.com/rancher/rancher/blob/5f98b0f28bc8a48366e9ced8f0a0fea3d826d90d/pkg/rbac/common.go#L46

It was introduced so that the monitoring app could create a service account and use rancher RBAC to manage its permissions: https://github.com/rancher/rancher/pull/19766

I'm guessing it used `norman:nocreate` in order to prevent users from using it, they just wanted the monitoring controller using it: https://github.com/rancher/rancher/pull/19766/files#diff-34975bed09fad198a231f28780f8194ba1129f987d56770a1cd79a5b00a15f85R329-R338

If the webhook forbids setting this field, that would actually break that controller.

I don't see a good reason to keep enforcing `nocreate` for this, just `noupdate` same as the other subject fields.

Seeing as this seems to only be needed for monitoring v1, maybe the new PRTB CRD could just drop it, and  delete that controller? But in the mean time I don't think the webhook needs to do anything special for it.

Reply:
Interesting 👀  That all sounds good to me.

Reply:
Only ported the noupdate in this PR. But I am curious about how the controller isn't already broken, since Norman with its struct tags must currently enforce nocreate...

Reply:
The controllers do not use Norman endpoints, only users.

Reply:
@cmurphy, @KevinJoiner - the webhook rejects the request if a PRTB doesn't have a user subject (user or user principal) AND doesn't have a group subject (group or group principal). What about ServiceAccount? It's also a kind of subject. This means there could be a user subject set AND a serviceAccount. Is this bad?

Reply:
Yes, ServiceAccount is a kind of subject and it is in conflict with User and Group as a subject. This needs to be fixed to validate that only one of the three types is set.

Reply:
good catch.